tests/python/unittest/test_container_structural_equal.py

```diff
@@ -0,0 +1,150 @@
+
+    return mismatch
+
+
+@pytest.mark.parametrize(
+    "a, b, expected_a_path, expected_b_path",
+    [
+        (
+            [1, 2, 3],
+            [1, 4, 3],
+            ObjectPath.root().array_index(1).attr("value"),
+            ObjectPath.root().array_index(1).attr("value"),
+        ),
+        (
+            [1, 2, 3],
+            [10, 2, 30],
+            ObjectPath.root().array_index(0).attr("value"),
+            ObjectPath.root().array_index(0).attr("value"),
+        ),
+        (
+            [1, 3, 4],
+            [1, 2, 3, 4],
+            ObjectPath.root().array_index(1).attr("value"),
+            ObjectPath.root().array_index(1).attr("value"),
+        ),
+        (
+            [1, 2, 3],
+            [1, 2, 3, 4],
+            ObjectPath.root().missing_array_element(3),
+            ObjectPath.root().array_index(3),
+        ),
+        (
+            [],
+            [1],
+            ObjectPath.root().missing_array_element(0),
+            ObjectPath.root().array_index(0),
+        ),
+    ],
+)
+def test_array_structural_mismatch(a, b, expected_a_path, expected_b_path):
+    a = tvm.runtime.convert(a)
+    b = tvm.runtime.convert(b)
+    a_path, b_path = get_first_mismatch_ensure_symmetry(a, b)
+    assert a_path == expected_a_path
+    assert b_path == expected_b_path
+
+
+@pytest.mark.parametrize(
+    "contents",
+    [
+        [],
+        [1],
+        [1, 2, 3],
+    ],
+)
+def test_array_structural_equal_to_self(contents):
+    a = tvm.runtime.convert(list(contents))
+    b = tvm.runtime.convert(list(contents))
+    assert get_first_mismatch_ensure_symmetry(a, b) is None
+
+
+@pytest.mark.parametrize(
+    "a, b, expected_a_path, expected_b_path",
+    [
+        (
+            dict(a=3, b=4),
+            dict(a=3, b=5),
+            ObjectPath.root().map_value("b").attr("value"),
+            ObjectPath.root().map_value("b").attr("value"),
+        ),
+        (
+            dict(a=3, b=4),
+            dict(a=3, b=4, c=5),
+            ObjectPath.root().missing_map_entry(),
+            ObjectPath.root().map_value("c"),
+        ),
+    ],
+)
+def test_string_map_structural_mismatch(a, b, expected_a_path, expected_b_path):
+    a = tvm.runtime.convert(a)
+    b = tvm.runtime.convert(b)
+    a_path, b_path = get_first_mismatch_ensure_symmetry(a, b)
+    assert a_path == expected_a_path
+    assert b_path == expected_b_path
+
+
+@pytest.mark.parametrize(
+    "contents",
+    [
+        dict(),
+        dict(a=1),
+        dict(a=3, b=4, c=5),
+    ],
+)
+def test_string_structural_equal_to_self(contents):
+    a = tvm.runtime.convert(dict(contents))
+    b = tvm.runtime.convert(dict(contents))
+    assert get_first_mismatch_ensure_symmetry(a, b) is None
+
+
+# The behavior of structural equality for maps with non-string keys is fairly specific
+# to IR variables because it assumes that map keys have been "mapped" using
+# `SEqualReducer::FreeVarEqualImpl()`. So we leave this case to TIR tests.
```

Comment:
nit: Let's append some epilogue to this file so that this single file could be executed independently using `python /path/to/file`.

```python
if __name__ == "__main__":
  tvm.testing.main()
```

Reply:
Done.

Seems unnecessary though -- why not simply run it like `pytest /path/to/file`? (And if you need the debugger then it's `gdb --args python -m pytest /path/to/file` -- seems simple enough)

Reply:
Yeah I personally don't think it necessary either...Just seeing other test files adding those lines